searchlib/src/vespa/searchlib/queryeval/multibitvectoriterator.cpp

```diff
@@ -1,61 +1,91 @@
 class MultiBitVectorIterator : public MultiBitVectorIteratorBase
 {
 public:
-    MultiBitVectorIterator(Children children) : MultiBitVectorIteratorBase(std::move(children)) { }
+    explicit MultiBitVectorIterator(Children children)
+        : MultiBitVectorIteratorBase(std::move(children)),
+          _update(),
+          _accel(IAccelrated::getAccelerator()),
+          _lastWords()
+    {
+        static_assert(sizeof(_lastWords) == 64, "Latswords should have 64 byte size");
+        memset(&_lastWords, 0, sizeof(_lastWords));
+    }
 protected:
     void updateLastValue(uint32_t docId);
     void strictSeek(uint32_t docId);
 private:
     void doSeek(uint32_t docId) override;
     Trinary is_strict() const override { return Trinary::False; }
     bool acceptExtraFilter() const override { return Update::isAnd(); }
-    Update                  _update;
+    Update              _update;
+    const IAccelrated & _accel;
+    alignas(64) Word    _lastWords[8];
 };
 
 template<typename Update>
 class MultiBitVectorIteratorStrict : public MultiBitVectorIterator<Update>
 {
 public:
-    MultiBitVectorIteratorStrict(MultiSearch::Children children) : MultiBitVectorIterator<Update>(std::move(children)) { }
+    explicit MultiBitVectorIteratorStrict(MultiSearch::Children  children)
+        : MultiBitVectorIterator<Update>(std::move(children))
+    { }
 private:
     void doSeek(uint32_t docId) override { this->strictSeek(docId); }
     Trinary is_strict() const override { return Trinary::True; }
 };
 
+struct And {
+    using Word = BitWord::Word;
+    void operator () (const IAccelrated & accel, size_t offset, const std::vector<std::pair<const void *, bool>> & src, Word *dest) {
+        accel.and64(offset*sizeof(uint64_t), src, dest);
+    }
+    static bool isAnd() { return true; }
+};
+
+struct Or {
+    using Word = BitWord::Word;
+    void operator () (const IAccelrated & accel, size_t offset, const std::vector<std::pair<const void *, bool>> & src, Word *dest) {
+        accel.or64(offset*sizeof(uint64_t), src, dest);
+    }
+    static bool isAnd() { return false; }
+};
+
 template<typename Update>
 void MultiBitVectorIterator<Update>::updateLastValue(uint32_t docId)
 {
     if (docId >= _lastMaxDocIdLimit) {
-        if (__builtin_expect(docId < _numDocs, true)) {
-            const uint32_t index(wordNum(docId));
-            _lastValue = _bvs[0][index];
-            for(uint32_t i(1); i < _bvs.size(); i++) {
-                _lastValue = _update(_lastValue, _bvs[i][index]);
-            }
-            _lastMaxDocIdLimit = (index + 1) * WordLen;
-        } else {
+        if (__builtin_expect(docId >= _numDocs, false)) {
             setAtEnd();
+            return;
+        }
+        const uint32_t index(wordNum(docId));
+        if (docId >= _lastMaxDocIdLimitRequireFetch) {
+            uint32_t baseIndex = index & ~(sizeof(_lastWords)/sizeof(Word) - 1);
+            _update(_accel, baseIndex, _bvs, _lastWords);
```

Comment:
I think the magic should be done outside the update call and the update call just forward to the appropriate function.

Reply:
Fixed